### PR TITLE
feat(coderd/database/dbtestutil): add ability to dump database on failure

### PR DESCRIPTION
**Pull request by johnstcn**
72dff7f18832bed585376830d333005dbb947842 included a provision to dump the contents of the database on failure, which actually seems really useful in a generic context.

- Extracted this to a more general function exposed in `dbtestutil`.
- This allows you to easily load data from a unit test and inspect the state at the time of test failure.
- I chose column insert format as this is the most portable and explicit, but COPY FROM STDIN is another option.
- I've chosen to not emit the dump to `t.Log` but this is also an option if running in CI. We just need to make sure to not leak any secrets.

### CODE REVIEW

File: coderd/database/dbtestutil/db.go [hunk 1]
```diff
@@ -110,3 +126,75 @@ func dbNameFromConnectionURL(t testing.TB, connectionURL string) string {
...
+	if !t.Failed() {
+		return
+	}
+	cwd, err := filepath.Abs(".")
+	if err != nil {
+		t.Errorf("dump on failure: cannot determine current working directory")
+		return
+	}
+	snakeCaseName := regexp.MustCompile("[^a-zA-Z0-9-_]+").ReplaceAllString(t.Name(), "_")
+	outPath := filepath.Join(cwd, snakeCaseName+".test.sql")
```
Comment (mafredri) [inline coderd/database/dbtestutil/db.go:140]: How about adding a random suffix to make sure we don't overwrite previous dumps?

**Review by mafredri — APPROVED** [review]
Nice work!

I would love for these to be uploaded as artifacts in CI, but if we're cautious about leaking secrets that might be an issue.

File: coderd/database/dbtestutil/db.go (hunk 1 above)
Comment (johnstcn) [reply coderd/database/dbtestutil/db.go:140]: Could add `YYYYmmddHHMMSS` so you know at a glance which one is the most recent?

File: coderd/database/dbtestutil/db.go [hunk 2]
```diff
@@ -110,3 +126,75 @@ func dbNameFromConnectionURL(t testing.TB, connectionURL string) string {
...
+		"--column-inserts",
+		"--no-comments",
+		"--no-privileges",
+		"--no-publication",
+		"--no-security-labels",
+		"--no-subscriptions",
+		"--no-tablespaces",
+		// "--no-unlogged-table-data", // some tables are unlogged and may contain data of interest
+		"--no-owner",
+		"--exclude-table=schema_migrations",
```
Comment (mtojek) [inline coderd/database/dbtestutil/db.go:171]: nit: should we skip more Coder tables by default?

**Review by mtojek — APPROVED** [review]
nit: you can also consider adding a short guide on how to import the dump into a local Coder deployment

File: coderd/database/dbtestutil/db.go (hunk 2 above)
Comment (johnstcn) [reply coderd/database/dbtestutil/db.go:171]: I'd rather err on the side of including them; schema_migrations is probably not that useful when inspecting unit test outputs though.

**Pull request merged**